prql-compiler/src/sql/gen_expr.rs

```diff
@@ -93,6 +93,11 @@ pub(super) fn translate_expr(expr: Expr, ctx: &mut Context) -> Result<sql_ast::E
                 Err(expr) => expr,
             };
 
+            let expr = match try_into_regex_function(expr, ctx)? {
+                Ok(between) => return Ok(between),
+                Err(expr) => expr,
+            };
```

Comment:
Yes unfortuanetely, this is checking each of the `try_into_x` one after another. It is actually my attempt at matching.

Good news is that it is necessary only for things that was specifically want to convert to `sql_parser`'s AST or that need special handling. If it could be implemented with an s-string, it can go into `std_impl.prql`.

So this PR could be done by just implementing `std.regex_search` in `std_impl.prql`, it we had some mechanism for changing the impl depending on the dialect.

Reply:
OK, I wonder whether we've over-abstracted this, and there's simpler code, even if more verbose and a bit less performant.

e.g. looking at this for a few minutes (but this only does `add`, would need to be extended to other binary ops, and doesn't handle an incorrect number of args)

```rust
        ExprKind::BuiltInFunction { name, .. } if name == "std.and" => {
            let op = BinaryOperator::Plus;
            let strength = op.binding_strength();
            let [left, right] = unpack(expr, STD_AND);
            let left = translate_operand(left, strength, !op.associates_left(), ctx)?;
            let right = translate_operand(right, strength, !op.associates_right(), ctx)?;
            sql_ast::Expr::BinaryOp { left, right, op }
        }
```

I don't have much confidence here though — raising since I'm still trying to go through the code and see if the perspective of someone worse at writing code can help! (And I won't pursue further unless requested)

Reply:
I also wonder if it's more performant. Even if it's not allocating, it's calling many functions on each resolution. 

(I realized this when I had https://github.com/PRQL/prql/pull/2458/files#diff-0a1f682019c767247e01027a9762cc4999669f393eb7971f095d81bea690163aR213-R219 before the `    let Some((decl, _)) = try_unpack(&expr, DECLS)? else {` — which panicked whenever mssql was running, even though it wasn't calling regex).

That said, I'm often surprised how cheap everything is relative to allocating.

But then I'm also often reminded of how cheap perf is relative to code complexity :)